File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/playlists/PlayListsInteractor.java
package com.github.anrimian.musicplayer.domain.interactors.playlists;

import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics;
import com.github.anrimian.musicplayer.domain.interactors.playlists.validators.PlayListNameValidator;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem;
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition;
import com.github.anrimian.musicplayer.domain.repositories.PlayListsRepository;
import com.github.anrimian.musicplayer.domain.repositories.UiStateRepository;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public class PlayListsInteractor {

    private final PlayListNameValidator nameValidator = new PlayListNameValidator();

    private final PlayListsRepository playListsRepository;
    private final UiStateRepository uiStateRepository;
    private final Analytics analytics;

    public PlayListsInteractor(PlayListsRepository playListsRepository,
                               UiStateRepository uiStateRepository,
                               Analytics analytics) {
        this.playListsRepository = playListsRepository;
        this.uiStateRepository = uiStateRepository;
        this.analytics = analytics;
    }

    public Observable<List<PlayList>> getPlayListsObservable() {
        return playListsRepository.getPlayListsObservable();
    }

    public Observable<PlayList> getPlayListObservable(long playListId) {
        return playListsRepository.getPlayListObservable(playListId);
    }

    public Observable<List<PlayListItem>> getCompositionsObservable(long playlistId) {
        return playListsRepository.getCompositionsObservable(playlistId);
    }

    public Single<PlayList> createPlayList(String name) {
        return nameValidator.validate(name)
                .flatMap(playListsRepository::createPlayList);
    }

    public Completable addCompositionsToPlayList(List<Composition> compositions,
                                                 PlayList playList,
                                                 int position) {
        return playListsRepository.addCompositionsToPlayList(compositions, playList, position);
    }

    public Completable addCompositionsToPlayList(List<Composition> compositions, PlayList playList) {
        return playListsRepository.addCompositionsToPlayList(compositions, playList, false);
    }

    public Completable deleteItemFromPlayList(PlayListItem playListItem, long playListId) {
        return playListsRepository.deleteItemFromPlayList(playListItem, playListId);
    }

    public Completable restoreDeletedPlaylistItem() {
        return playListsRepository.restoreDeletedPlaylistItem();
    }

    public Completable deletePlayList(long playListId) {
        return playListsRepository.deletePlayList(playListId);
    }

    public Completable moveItemInPlayList(PlayList playList, int from, int to) {
        return playListsRepository.moveItemInPlayList(playList, from, to)
                .doOnError(analytics::processNonFatalError);
    }

    public Completable updatePlayListName(long playListId, String name) {
        return nameValidator.validate(name)
                .flatMapCompletable(o -> playListsRepository.updatePlayListName(playListId, name));
    }

    public void setSelectedPlayListScreen(long playListId) {
        uiStateRepository.setSelectedPlayListScreenId(playListId);
    }

    public void saveListPosition(ListPosition listPosition) {
        uiStateRepository.savePlaylistsPosition(listPosition);
    }

    public ListPosition getSavedListPosition() {
        return uiStateRepository.getSavedPlaylistsPosition();
    }

    public void saveItemsListPosition(long playListId, ListPosition listPosition) {
        uiStateRepository.savePlaylistsListPosition(playListId, listPosition);
    }

    public ListPosition getSavedItemsListPosition(long playListId) {
        return uiStateRepository.getSavedPlaylistListPosition(playListId);
    }
}

File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/playlists/PlayListsInteractor.kt
package com.github.anrimian.musicplayer.domain.interactors.playlists

import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics
import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor
import com.github.anrimian.musicplayer.domain.interactors.playlists.validators.PlayListFileNameValidator
import com.github.anrimian.musicplayer.domain.interactors.playlists.validators.PlayListNameValidator
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.folders.FileReference
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition
import com.github.anrimian.musicplayer.domain.repositories.PlayListsRepository
import com.github.anrimian.musicplayer.domain.repositories.SettingsRepository
import com.github.anrimian.musicplayer.domain.repositories.UiStateRepository
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single

class PlayListsInteractor(
    private val playerInteractor: LibraryPlayerInteractor,
    private val playListsRepository: PlayListsRepository,
    private val settingsRepository: SettingsRepository,
    private val uiStateRepository: UiStateRepository,
    private val analytics: Analytics
) {

    private val nameValidator = PlayListNameValidator()

    fun getPlayListsObservable(searchQuery: String? = null): Observable<List<PlayList>> {
        return playListsRepository.getPlayListsObservable(searchQuery)
    }

    fun getPlayListObservable(playListId: Long): Observable<PlayList> {
        return playListsRepository.getPlayListObservable(playListId)
    }

    fun getCompositionsObservable(
        playlistId: Long,
        searchText: String?
    ): Observable<List<PlayListItem>> {
        return playListsRepository.getCompositionsObservable(playlistId, searchText)
    }

    fun createPlayList(name: String): Single<PlayList> {
        return nameValidator.validate(PlayListFileNameValidator.normalizePlayListName(name))
            .flatMap(playListsRepository::createPlayList)
    }

    fun addCompositionsToPlayList(
        compositions: List<Composition>,
        playList: PlayList,
        checkForDuplicates: Boolean,
        ignoreDuplicates: Boolean
    ): Single<List<Composition>> {
        val duplicateCheck = checkForDuplicates && settingsRepository.isPlaylistDuplicateCheckEnabled
        return playListsRepository.addCompositionsToPlayList(
            compositions,
            playList,
            duplicateCheck,
            ignoreDuplicates
        )
    }

    fun deleteItemFromPlayList(playListItem: PlayListItem, playListId: Long): Completable {
        return playListsRepository.deleteItemFromPlayList(playListItem, playListId)
    }

    fun restoreDeletedPlaylistItem(): Completable {
        return playListsRepository.restoreDeletedPlaylistItem()
    }

    fun deletePlayList(playListId: Long): Completable {
        return playListsRepository.deletePlayList(playListId)
    }

    fun deletePlayLists(playlists: Collection<PlayList>): Completable {
        return Observable.fromIterable(playlists)
            .flatMapCompletable { playlist -> playListsRepository.deletePlayList(playlist.id) }
    }

    fun moveItemInPlayList(playListId: Long, from: Int, to: Int): Completable {
        return playListsRepository.moveItemInPlayList(playListId, from, to)
            .doOnError(analytics::processNonFatalError)
    }

    fun updatePlayListName(playListId: Long, name: String): Completable {
        return nameValidator.validate(PlayListFileNameValidator.normalizePlayListName(name))
            .flatMapCompletable { playListsRepository.updatePlayListName(playListId, name) }
    }

    fun setSelectedPlayListScreen(playListId: Long) {
        uiStateRepository.selectedPlayListScreenId = playListId
    }

    fun saveListPosition(listPosition: ListPosition?) {
        uiStateRepository.savePlaylistsPosition(listPosition)
    }

    fun getSavedListPosition(): ListPosition? = uiStateRepository.savedPlaylistsPosition

    fun saveItemsListPosition(playListId: Long, listPosition: ListPosition) {
        uiStateRepository.savePlaylistsListPosition(playListId, listPosition)
    }

    fun getSavedItemsListPosition(playListId: Long): ListPosition? {
        return uiStateRepository.getSavedPlaylistListPosition(playListId)
    }

    fun isPlaylistDuplicateCheckEnabled() = settingsRepository.isPlaylistDuplicateCheckEnabled

    fun setPlaylistDuplicateCheckEnabled(isEnabled: Boolean) {
        settingsRepository.isPlaylistDuplicateCheckEnabled = isEnabled
    }

    fun exportPlaylistsToFolder(playlists: List<PlayList>, folder: FileReference): Completable {
        return playListsRepository.exportPlaylistsToFolder(playlists, folder)
    }

    fun importPlaylistFile(file: FileReference, overwriteExisting: Boolean): Single<Long> {
        return playListsRepository.importPlaylistFile(file, overwriteExisting)
    }

    fun startPlaying(playlists: List<PlayList>): Completable {
        return playListsRepository.getCompositionIdsInPlaylists(playlists)
            .flatMapCompletable(playerInteractor::setQueueAndPlay)
    }

    fun getCompositionsByPlaylistsIds(playlistIds: LongArray): Single<List<Composition>> {
        return playListsRepository.getCompositionsByPlaylistsIds(playlistIds.asIterable())
    }

    fun getCompositionsInPlaylists(playlists: List<PlayList>): Single<List<Composition>> {
        return playListsRepository.getCompositionsInPlaylists(playlists)
    }

}